### ISSUE TITLE
Having trouble using the command console with MyComponent

### ISSUE THREAD

**Issue by mooror** [BobRay/MyComponent#29]
Hey there BobRay, First off I want to thank you so much for this wonderful extra. It has (and I think will continue to be) an amazing asset. Second I was reading your documentation on the extra on your website when I came across this line : "The utilities can be run from the command line, or from inside any good code editor. ". This got me extremely exited so I tried finding any documentation I could on this functionality of MyComponent. This is where I ran into issues as I couldn't really find to much info on this subject. So I just winged it and tried running 
```
php bootstrap.php
```
from the console in the /core/components/mycomponent/elements/snippets/ folder of modx. This resulted in this error: 
```
[bootstrap.php] Could not find build.config.php
``` 
 Perhaps I am not executing this file correctly, or something else but I just thought I would get your expert advice before trying anything stupid.

Thanks in advance,
Mooror

**Comment by mooror** [comment]
Also as a side note I am using putty to connect to my server and run the above command.

**Comment by mooror** [comment]
Oh wait I cant bootstrap anything until I create the project first. So now the question is what is the command-line equivalent of the new project button in the ui?

**Comment by BobRay** [comment]
The easiest thing is to create the new project in the MODX Manager, edit the config file (it will tell you where it is), then you should be able to run bootstrap and the other utilities. You  may need to have the _build directory in your include path (or hardcode the path to the build.config file in the snippets).

I seldom use the command line. I run and debug the snippets in PhpStorm or run MyComponent from the Manager form.

**Comment by mooror** [comment]
I was scared of that. See we are trying to make our system semi automatic. Is there any way to manually setup a new project? like copy over the folder structure and duplicate the config and rename it. Or is is a lot more involved the that?

**Comment by BobRay** [comment]
Using my method, you get the config file from the example project, modified for your project name. It comes from a Tpl chunk (example.config.php), but you can create your own Tpl chunk for it to use. IIRC, if you call your chunk myexample.config.php, it will be used automatically.

This discussion really belongs in the MODX Forums, since it's not really a bug or feature request.

**Comment by mooror** [comment]
Hello again Bob. I would definitely use the form unfortunately my account was deactivated for reasons unknown to me and I am still trying to deal with that. I was able to get the project creation handled, however I'm still not sure how to bootstrap my project from the console. For instance my project is called siteleasecore. So there has been a siteleasecore.config.php file created for it in the mycomponent/_build/config/ folder. Now how do I bootstrap that project/file via the command line? 

Thanks in advance,
Mooror

P.s. Thank you for all your replies and timely help.

**Comment by BobRay** [comment]
It the project config file exists and is in the right place (and doesn't contain any PHP errors), you should be able to just run bootstrap. Be sure that the mycomponent/_build/config/current.project.php file contains the name of your project.

You may have to modify your include path.

Be sure error reporting is on so you'll see problems that occur. Bootstrap takes a while to run, so be patient.

**Comment by mooror** [comment]
Thanks Bob for your help. Its been very amazing so far. I got myproject created and bootstrapped properly now from the command line(using the current.project.php for the bootstrapping), but now I'm wondering how I can edit the and add new elements to my bootstrapped project. I was reading your documentation looking for where the bootstrapped elements are stored but I couldn't find anything on it. Are all the elements for all the bootstraped projects stored in the actual modx install? Or are they stored in static files somewhere (I saw the mycomponent/elements/ folder, not sure if that's what I'm looking for).

Thanks in advance,
Mooror

**Comment by mooror** [comment]
So I was lloking around and found the "Trying to use MyComponent" thread on modx. I read the flowchart as well as some of the comments and guessed that the "stub" files are supposed to go in the assets/mycomponents/myProject/core/ folder. So I want to that path in my modx install and descovered that the assets/mycomponents folder only contains  the _build folder. So next I tried using the ui to see if it would do anything different and found that it created assets/mycomponents/myProject/core/ just fine with all of the elements folders in there. I'm not sure if this issue is repeatable, but I'm going to be looking into a possible permissions error before making any rash conclusions. Ill document my findings here for future reference.

**Comment by mooror** [comment]
Hmm I was calling ```php bootstrap.php```  from inside a script. The output my have not been logging to the scripts console. When I ran the command in the commandline over putty I got this error: `[bootstrap.php] Could not find build.config.php`

**Comment by mooror** [comment]
I checked the bootstrap.php file and found this code:
```
if (!defined('MODX_CORE_PATH')) {
    $path1 = dirname(dirname(dirname(dirname(dirname(dirname(__FILE__)))))) . '/_build/build.config.php';
    if (file_exists($path1)) {
        include $path1;
    } else {
        $path2 = dirname(dirname(dirname(__FILE__))) . '/_build/build.config.php';
        if (file_exists($path2)) {
            include($path2);
        }
    }
    if (!defined('MODX_CORE_PATH')) {
        session_write_close();
        die('[bootstrap.php] Could not find build.config.php');
    }
```
From what i can tell this code is testing two paths to the build.config.php to see if they exist (in other words that the file exists at the path specified). If the both don't exist then die and error and end the current session and store session data. Both the paths if I understand correctly are somehow using the files own path as a base (if I'm correctly understanding what dirname(__FILE__) does) and then looking for a _build/build.config.php file under that. However when I look in my own core/components/mycomponent/_build/ folder I see only the config and utilities folders. I assume I am looking in the wrong _build folder, but either way the ui works when running the script directly from the commandline doesn't seem to. So either there is a step run by the ui that is not run by the commanline bootstrap.php file or I am totally missing something.

**Comment by mooror** [comment]
I also get these errors in the ui:
```
Parse error: Missing operand in file 'home.js' on line 31
Parse error: Missing operand in file 'chunk.grid.js' on line 26
Parse error: Missing operand in file 'snippet.grid.js' on line 26
Parse error: Missing operand in file 'home.panel.js' on line 27
```

**Comment by mooror** [comment]
I am also getting some very strange behavior. If i create a new project and don't change the config file before bootstrapping I get a folder in the assets/mycomponents/ folder. But if i do change the config file then the contents that are usually stored under a folder in assets/mycomponents/ are stored directly in the assets/mycomponents/.

**Comment by BobRay** [comment]
Wow, so many questions.

First, the MC utilities will technically run from the command line, but they weren't really designed with that in mind. They were designed to run from the UI, or in a code editor (mainly so I could debug them). They might work from the command line if you navigate to the directory with the utilities first.

The dual search for the core is similar. It looks where the relative location of the config.core.php file would be in my development environment, if that fails, it looks where a typical user would have it installed. It will work if either one is found.

To edit objects, you can edit them in MODX and use ExportObjects to update the files, or you can edit the files and use ImportObjects to update the objects in MODX.

For new objects, the best way is to update the project config file (the one in the MyComponent directory) and run Bootstrap. That will create the objects in MODX and the files on the disk (for those objects that have content and take files). 

I assume that you've seen this: ( http://bobsguides.com/mycomponent-tutorial.html ) but it might be worth re-reading.

**Comment by mooror** [comment]
Thank you for your help so far. I think I understand the workflow better now. Just one more thing that I would like to clear up. I tried what you suggested and went to the mycomponent/elements/snippets/ folder and then tried running the bootstrap utility but unfortunately I still get this error `[bootstrap.php] Could not find build.config.php`. I already made sure that a) I used the ui to start the project (the name of which is sitename). and b) that the project was the currentproject (I even checked the current.project.php file to make sure).  But I think my workspace might be lacking something that you have when you use your commandline. The reason I say this is right after I tried to run the bootstrap via commandline (and got the error) I tried bootstrapping the same project through the ui and it worked fine. So I have to ask,when you have used the commandline utilities (for debugging or anything else) did you have to relocate/copy the build.config.php file somewhere or change the path that the utilities look for that file? It seems that inside modx the utility (snippet) can find the build.config.php file fine however outside modx it cannot. That's my guess.

Thank you very much,
Mooror

p.s. I am going to try to modify the path to the build.config.php file in the bootstrap.php snippet (located at mycomponent/elements/snippets/)

**Comment by mooror** [comment]
Come to think of it where is the "build.config.php" normally stored? I cannot find it in my install.

**Comment by mooror** [comment]
Okay so I replaced this line of code (in the bootstrap.php file):
```
$path1 = dirname(dirname(dirname(dirname(dirname(dirname(__FILE__)))))) . '/_build/build.config.php';
```
With this code (feel free to warn me if this is a bad idea): 
```
if (php_sapi_name() != 'cli') {
      $path1 = dirname(dirname(dirname(dirname(dirname(dirname(__FILE__)))))) . '/_build/build.config.php';
    }
    else {
      $path1 = '/*my hard coded path*/assets/mycomponents/_build/build.config.php';
    }
```
Then I reran the bootstrap.php file and got a different error (or rather errors) then before. Here they are:
```
[2015-12-09 16:53:35] (ERROR @ /*my hard coded path*/core/components/mycomponent/elements/snippets/bootstrap.php : 64) PHP warning: require_once(/*my hard coded path*/core/components/mycomponent/model/mycomponent/mycomponentproject.class.php): failed to open stream: No such file or directory

Fatal error: require_once(): Failed opening required '/*my hard coded path*/core/components/mycomponent/model/mycomponent/mycomponentproject.class.php' (include_path='.:/usr/lib/php:/usr/local/lib/php') in /*my hard coded path*/core/components/mycomponent/elements/snippets/bootstrap.php on line 64
```
I'm not 100% sure why this is happening but I did find this in my assets/mycomponents/_build/build.config.php file:
```
 /* next line require only for operating outside of MODX
    Create a file outside of the MODX web root with this content:
     <?php
     $fields = array();
     $fields['username'] = 'your_MODX_username';
     $fields['password'] = 'your_MODX_password';
     $fields['login_context'] = 'mgr';
     $fields['add_contexts'] = 'mgr';

     Put the full path to the file in the next line
  */
```
Do you think this might be what I need to do to get permissions to the modx install from the bootstrap.php file?

Thanks,
Mooror

**Comment by BobRay** [comment]
That's very strange. That assets/mycomponents/_build/build.config.php file
doesn't exist in my install, and you shouldn't need credentials to run any
of the utilities in CLI mode.

The build.config.php file should be in the
assets/mycomponents/yourcomponentname/_build directory.

It looks like this:

/* Define the MODX path constants necessary for connecting to your
core and other directories.
 * If you have not moved the core, the current values should work.
 * In some cases, you may have to hard-code the full paths */
if (!defined('MODX_CORE_PATH')) {
    define('MODX_CORE_PATH',
dirname(dirname(dirname(dirname(dirname(__FILE__))))) . '/core/');
    define('MODX_BASE_PATH',
dirname(dirname(dirname(dirname(dirname(__FILE__))))) . '/');
    define('MODX_MANAGER_PATH', MODX_BASE_PATH . 'manager/');
    define('MODX_CONNECTORS_PATH', MODX_BASE_PATH . 'connectors/');
    define('MODX_ASSETS_PATH', MODX_BASE_PATH . 'assets/');
}

/* not used -- here to prevent E_NOTICE warnings */
if (!defined('MODX_BASE_URL')) {
    define('MODX_BASE_URL', 'http://localhost/addons/');
    define('MODX_MANAGER_URL', 'http://localhost/addons/manager/');
    define('MODX_ASSETS_URL', 'http://localhost/addons/assets/');
    define('MODX_CONNECTORS_URL', 'http://localhost/addons/connectors/');
}

Bob

**Comment by mooror** [comment]
Hey there Bob, thanks for the answer. I am happy to announce that last night I got it working just fine from the commandline. However in light of what you have just told me, I think that my solution to the problem might need a small tweak. Even so I am glad to get it working from the commandline.

Mooror
p.s. I will close this issue once I integrate the new information you have given me into my solution and post the solution here.

**Comment by mooror** [comment]
Right, I understand now.So here is the recap and solution.When I got the commandline working yesterday I went looking for that build.config.php file that the bootstrap.php utility was looking for. I mistakenly thought it was that build file at assets/mycomponents/_build/build.config which was probably a leftover piece from one of my failed package bootstrap attempts. So when I had hard coded in the path to that file (into the bootstrap.php file) and reran the script it gave me these errors:
```
[2015-12-09 16:53:35] (ERROR @ /*my hard coded path*/core/components/mycomponent/elements/snippets/bootstrap.php : 64) PHP warning: require_once(/*my hard coded path*/core/components/mycomponent/model/mycomponent/mycomponentproject.class.php): failed to open stream: No such file or directory

Fatal error: require_once(): Failed opening required '/*my hard coded path*/core/components/mycomponent/model/mycomponent/mycomponentproject.class.php' (include_path='.:/usr/lib/php:/usr/local/lib/php') in /*my hard coded path*/core/components/mycomponent/elements/snippets/bootstrap.php on line 64
```
I looked at the paths in that error more closely and found that the path to my modx core was incorrect. It was missing one folder. So doing some testing I discovered that changing this line (in the build.config.php file):
```
define('MODX_CORE_PATH', dirname(dirname(dirname(dirname(dirname(__FILE__))))) . '/core/');
```
to this:
```
define('MODX_CORE_PATH', dirname(dirname(dirname(dirname(__FILE__)))) . '/core/');
```
fixed those errors and allowed me to run the utilities from the commandline. But today when you informed me that the `assets/mycomponents/_build/build.config.php` path didn't exist in your install and that it should exist here `assets/mycomponents/yourcomponentname/_build` everything made sense. You had designed the build.config.php file to be used from within the above path by using the `dirname(__FILE__)` to get the directory structures using **the files current location**. Therefor if the files location is different then the paths will be off and the script will error out. To confirm this I changed this code that I added earlier (bootstrap.php):
```
if (php_sapi_name() != 'cli') {
      $path1 = dirname(dirname(dirname(dirname(dirname(dirname(__FILE__)))))) . '/_build/build.config.php';
    }
    else {
      $path1 = '/*my hard coded path*/assets/mycomponents/_build/build.config.php';
    }
```
to use the `assets/mycomponents/yourcomponentname/_build` path like you suggested:
```
 if (php_sapi_name() != 'cli') {
      $path1 = dirname(dirname(dirname(dirname(dirname(dirname(__FILE__)))))) . '/_build/build.config.php';
    }
    else {
      $path1 =  dirname(dirname(dirname(dirname(dirname(dirname(__FILE__)))))) . '/assets/mycomponents/example/_build/build.config.php';
    }
```
and the utilities worked **without me having to edit the build.config.php file** like I had done last night.

So I guess this fixes the problem I am having and though it would be nice not to have to hard code the path I think it should work okay. Thank you for the wonderful conversation, it has enabled me to learn and grow in my php programming and for that I am very grateful. Hopefully we will cross paths again soon, but next time for a more casual conversation :D. Cheers!

Mooror
P.S. I forked this extra and will be working on creating a commandline version of the create project utility.
